lib/engine/game/g_1824/step/buy_sell_par_exchange_shares.rb

```diff
@@ -63,29 +82,89 @@ def process_buy_shares(action)
           end
 
           def buyable_items(entity)
-            return [] unless @game.coal_railway_exchangable?
+            items = []
+
+            if @game.coal_railway_exchangable?
+              @game.companies.select { |c| @game.coal_railway?(c) && c.owner == entity }.each do |c|
+                items << Item.new(description: c.id, cost: 0)
+              end
+            end
```

Comment:
I think this can be done in one iteration, instead of using both `select` and `each`, something like this might work:

```suggestion
            if @game.coal_railway_exchangable?
              items = @game.companies.each_with_object([]) do |c, i|
                 i << Item.new(description: c.id, cost: 0) if @game.coal_railway?(c) && c.owner == entity
              end
            end
```